File: 02-Linked-list/insertion/begin.c
#include<stdio.h>

typedef struct Node {
    int data;
    struct Node *next;
} Node;

Node *insertATStart(Node *head, int data) {
    static Node newNode;

    newNode.data = data;
    newNode.next = head;
    // new head
    return &newNode;
}

void printList(Node *head) {
    Node *curr = head;

    while(curr != NULL) {
        printf("%d", curr->data);

        if(curr->next != NULL) printf("->");

        // move to next node
        curr = curr->next;
    }

    printf("\n");
}

int main() {
    Node head = {10, NULL};
    Node node1 = {20, NULL};
    Node node2 = {30, NULL};
    Node node3 = {40, NULL};

    // set pointers
    head.next = &node1;
    node1.next = &node2;
    node2.next = &node3;

    //list before insert
    printList(&head);

    // insert at start
    Node *ptr = insertATStart(&head, 100);

    // list after insert
    printList(ptr);


    return 0;
}

File: 02-Linked-list/insertion/begin.js
class Node {
    data;
    next 

    constructor(data) {
        this.data = data;
        this.next = null
    }

    static insertAtTheBeginning(currentHead, data) {
        const newHead = new Node(data)

        newHead.next = currentHead
        return newHead

    }

    static printLinkedList(head) {
        let node = head;
        while (node !== null) {
          process.stdout.write(`${node.data}`);
          if (node.next) process.stdout.write(`--->`);

          node = node.next;
        }

        process.stdout.end();
        console.log();

    } 
}

let head = new Node(10)
head.next = new Node(20);
head.next.next = new Node(30);
head.next.next.next = new Node(50);


head = Node.insertAtTheBeginning(head, 100);
Node.printLinkedList(head)

